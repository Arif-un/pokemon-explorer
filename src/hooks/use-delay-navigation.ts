import { useState } from 'react'

import { useRouter } from '@tanstack/react-router'

interface NextLocation {
  to: string
  replace: boolean
}

export function useDelayedNavigation() {
  const router = useRouter()
  const [isNavigating, setIsNavigating] = useState(false)
  const [nextLocation, setNextLocation] = useState<NextLocation | null>(null)

  const navigate = (to: string, { replace = false } = {}) => {
    setIsNavigating(true)
    setNextLocation({ to, replace })
  }

  const finishNavigation = () => {
    if (nextLocation) {
      const { to, replace } = nextLocation
      router.navigate({ to, replace, search: prv => ({ ...prv }) })
      setIsNavigating(false)
      setNextLocation(null)
    }
  }

  return { isNavigating, navigate, finishNavigation }
}
