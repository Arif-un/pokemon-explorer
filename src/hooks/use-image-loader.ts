import { useEffect, useState } from 'react'

interface ImageState {
  src: string
  isLoading: boolean
  error: Event | string | null
}

const imageCache: Record<string, boolean> = {}

export function useImageLoader(src: string, fallbackSrc: string): ImageState {
  const [imageState, setImageState] = useState<ImageState>({
    src: fallbackSrc,
    isLoading: true,
    error: null
  })

  useEffect(() => {
    if (imageCache[src]) {
      setImageState({
        src: src,
        isLoading: false,
        error: null
      })
      return
    }

    const img = new Image()
    img.src = src

    img.onload = () => {
      imageCache[src] = true

      setImageState({
        src: src,
        isLoading: false,
        error: null
      })
    }

    img.onerror = error => {
      setImageState({
        src: fallbackSrc,
        isLoading: false,
        error
      })
    }

    return () => {
      img.onload = null
      img.onerror = null
    }
  }, [src, fallbackSrc])

  return imageState
}
