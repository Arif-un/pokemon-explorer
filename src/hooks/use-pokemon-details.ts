import { useQuery } from '@tanstack/react-query'

import { fetchPokemonDetails } from '@/services/pokemon-api'
import { type PokemonDetailsResponse } from '@/types/Pokemon'

export const usePokemonDetails = <SelectedData = PokemonDetailsResponse>(
  name: string,
  select: (data: PokemonDetailsResponse) => SelectedData
) => {
  const { data, isLoading, isError, error } = useQuery<PokemonDetailsResponse, Error, SelectedData>({
    queryKey: ['pokemon', name],
    queryFn: () => fetchPokemonDetails(name),
    select
  })

  return { data: data as SelectedData, isLoading, isError, error }
}
