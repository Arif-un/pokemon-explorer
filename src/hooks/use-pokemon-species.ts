import { useQuery } from '@tanstack/react-query'

import { fetchPokemonSpecies } from '@/services/pokemon-api'

export const usePokemonSpecies = (name: string) => {
  const { data, isLoading, isError, error } = useQuery({
    queryKey: ['pokemon-color', name],
    queryFn: () => fetchPokemonSpecies(name),
    staleTime: 5 * 60 * 1000
  })

  return {
    pokemonColor: data?.color.name,
    isLoading,
    isPokemonSpeciesError: isError,
    pokemonSpeciesError: error
  }
}
