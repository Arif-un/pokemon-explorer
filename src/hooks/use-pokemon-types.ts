import { useQuery } from '@tanstack/react-query'

import { fetchPokemonTypes } from '@/services/pokemon-api-graphql'
import { DEFAULT_QUERY_STALE_TIME } from '@/utils/constants'

export function usePokemonTypes() {
  const { data, isLoading, isError, error } = useQuery({
    queryKey: ['pokemon-types'],
    queryFn: fetchPokemonTypes,
    staleTime: DEFAULT_QUERY_STALE_TIME
  })

  return {
    types: data?.pokemon_v2_type.sort((a, b) => a.name.localeCompare(b.name)),
    isLoading,
    isError,
    error
  }
}
