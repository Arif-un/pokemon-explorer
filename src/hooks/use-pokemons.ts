import { useQuery } from '@tanstack/react-query'

import { fetchPokemonWithFilters } from '@/services/pokemon-api-graphql'
import {
  DEFAULT_POKEMON_FETCH_LIMIT,
  DEFAULT_POKEMON_FETCH_OFFSET,
  DEFAULT_QUERY_STALE_TIME
} from '@/utils/constants'

interface UsePokemonsParams {
  name?: string
  limit?: number
  offset?: number
  types?: string
  sort?: string
}

export const usePokemons = ({
  name,
  limit = DEFAULT_POKEMON_FETCH_LIMIT,
  offset = DEFAULT_POKEMON_FETCH_OFFSET,
  types,
  sort
}: UsePokemonsParams) => {
  const { data, isLoading, isError, error, isFetching } = useQuery({
    queryKey: ['pokemon', name, limit, offset, types, sort],
    queryFn: () => fetchPokemonWithFilters({ name, limit, offset, types, sort }),
    staleTime: DEFAULT_QUERY_STALE_TIME,
    placeholderData: previousData => previousData
  })

  return {
    pokemons: data?.pokemons,
    totalItems: data?.pokemon_aggregate.aggregate.count,
    isLoading,
    isError,
    error,
    isFetching
  }
}
