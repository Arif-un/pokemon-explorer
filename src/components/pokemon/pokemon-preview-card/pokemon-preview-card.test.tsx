import { type HTMLProps } from 'react'

import { render, screen } from '@testing-library/react'
import { describe, expect, it, vi } from 'vitest'

import { usePokemonDetails } from '@/hooks/use-pokemon-details'

import PokemonPreviewCard from './pokemon-preview-card'

vi.mock('@tanstack/react-router', () => ({
  Link: ({ children, ...props }: HTMLProps<HTMLAnchorElement>) => <a {...props}>{children}</a>
}))

vi.mock('@/hooks/use-pokemon-details', () => ({
  usePokemonDetails: vi.fn()
}))

const mockPokemonDetails = 'https://example.com/shiny.png'

describe('PokemonPreviewCard', () => {
  it('renders correctly while loading', () => {
    vi.mocked(usePokemonDetails).mockReturnValue({
      data: undefined,
      isLoading: true,
      isError: false,
      error: null
    })
    const { asFragment } = render(<PokemonPreviewCard name="Pikachu" />)

    expect(asFragment()).toMatchSnapshot()

    const loadingImage = screen.getByAltText('pokemon loading image')
    expect(loadingImage).toHaveAttribute('src', expect.stringContaining('pokemon-loader.svg'))
  })

  it('renders correctly with pokemon data', () => {
    vi.mocked(usePokemonDetails).mockReturnValue({
      data: mockPokemonDetails,
      isLoading: false,
      isError: false,
      error: null
    })

    const { asFragment } = render(<PokemonPreviewCard name="Charizard" />)

    expect(asFragment()).toMatchSnapshot()

    const pokemonImage = screen.queryByAltText('Charizard image')
    expect(pokemonImage).toHaveAttribute('src', 'https://example.com/shiny.png')

    const pokemonName = screen.getByText('Charizard')
    expect(pokemonName).toBeInTheDocument()
  })

  it('renders correctly with long name', () => {
    vi.mocked(usePokemonDetails).mockReturnValue({
      data: mockPokemonDetails,
      isLoading: false,
      isError: false,
      error: null
    })

    const longName = 'PikachuPikachuPikachuPikachu'
    const { asFragment } = render(<PokemonPreviewCard name={longName} />)

    expect(asFragment()).toMatchSnapshot()

    const pokemonName = screen.getByText(longName)
    expect(pokemonName).toHaveClass('text-base')
  })
})
