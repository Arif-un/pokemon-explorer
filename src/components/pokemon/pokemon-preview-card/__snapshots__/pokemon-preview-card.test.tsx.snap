// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PokemonPreviewCard > renders correctly while loading 1`] = `
<DocumentFragment>
  <a
    aria-label="View details of Pikachu"
    class="inline-block focus-visible:outline outline-yellow-500 outline-2 focus-visible:outline-offset-8 outline-offset-0  rounded-3xl transition-offset duration-200 motion-reduce:duration-0"
    to="pokemon-details/Pikachu"
  >
    <div
      class="inner-glow bg-slate-800 inline-flex flex-col justify-center group cursor-pointer  dark:bg-slate-900 p-2 dark:text-slate-50 rounded-3xl dark:border-slate-800 border"
    >
      <div
        class="parallax-effect-glare-scale "
      >
        <figure
          class="popup-element flex items-center justify-center"
        >
          <div
            class="relative w-60 h-60 -mt-16"
          >
            <img
              alt="pokemon loading image"
              class="w-40 m-auto"
              src="/src/assets/pokemon-loader.svg"
            />
          </div>
        </figure>
        <div
          class="flex items-center justify-center gap-2"
        >
          <h3
            class="text-slate-300 group-hover:text-slate-50 dark:text-slate-400 dark:group-hover:text-slate-50 group-hover:underline capitalize text-center font-semibold my-5 text-xl"
          >
            Pikachu
          </h3>
          <svg
            aria-hidden="true"
            aria-label="navigate icon"
            class="lucide lucide-arrow-up-right opacity-0 size-0 scale-0 group-hover:opacity-100 group-hover:scale-100 group-hover:size-auto transition-transform text-slate-50 dark:text-slate-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 7h10v10"
            />
            <path
              d="M7 17 17 7"
            />
          </svg>
        </div>
        <div
          class="glare-wrapper"
          style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; overflow: hidden; border-radius: 24px; pointer-events: none;"
        >
          <div
            class="glare"
            style="position: absolute; top: 50%; left: 50%; transform-origin: 0% 0%; pointer-events: none; width: 0px; height: 0px;"
          />
        </div>
      </div>
    </div>
  </a>
</DocumentFragment>
`;

exports[`PokemonPreviewCard > renders correctly with long name 1`] = `
<DocumentFragment>
  <a
    aria-label="View details of PikachuPikachuPikachuPikachu"
    class="inline-block focus-visible:outline outline-yellow-500 outline-2 focus-visible:outline-offset-8 outline-offset-0  rounded-3xl transition-offset duration-200 motion-reduce:duration-0"
    to="pokemon-details/PikachuPikachuPikachuPikachu"
  >
    <div
      class="inner-glow bg-slate-800 inline-flex flex-col justify-center group cursor-pointer  dark:bg-slate-900 p-2 dark:text-slate-50 rounded-3xl dark:border-slate-800 border"
    >
      <div
        class="parallax-effect-glare-scale "
      >
        <figure
          class="popup-element flex items-center justify-center"
        >
          <div
            class="relative w-60 h-60 -mt-16"
          >
            <img
              alt="PikachuPikachuPikachuPikachu image-backdrop"
              class="blur-2xl w-52 absolute opacity-50 group-hover:opacity-90 m-auto inset-0 transition-opacity"
              height="240"
              loading="lazy"
              src="https://example.com/shiny.png"
              width="240"
            />
            <img
              alt="PikachuPikachuPikachuPikachu image"
              class="w-52 m-auto z-10 relative"
              height="240"
              loading="lazy"
              src="https://example.com/shiny.png"
              width="240"
            />
          </div>
        </figure>
        <div
          class="flex items-center justify-center gap-2"
        >
          <h3
            class="text-slate-300 group-hover:text-slate-50 dark:text-slate-400 dark:group-hover:text-slate-50 group-hover:underline capitalize text-center font-semibold my-5 text-base"
          >
            PikachuPikachuPikachuPikachu
          </h3>
          <svg
            aria-hidden="true"
            aria-label="navigate icon"
            class="lucide lucide-arrow-up-right opacity-0 size-0 scale-0 group-hover:opacity-100 group-hover:scale-100 group-hover:size-auto transition-transform text-slate-50 dark:text-slate-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 7h10v10"
            />
            <path
              d="M7 17 17 7"
            />
          </svg>
        </div>
        <div
          class="glare-wrapper"
          style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; overflow: hidden; border-radius: 24px; pointer-events: none;"
        >
          <div
            class="glare"
            style="position: absolute; top: 50%; left: 50%; transform-origin: 0% 0%; pointer-events: none; width: 0px; height: 0px;"
          />
        </div>
      </div>
    </div>
  </a>
</DocumentFragment>
`;

exports[`PokemonPreviewCard > renders correctly with pokemon data 1`] = `
<DocumentFragment>
  <a
    aria-label="View details of Charizard"
    class="inline-block focus-visible:outline outline-yellow-500 outline-2 focus-visible:outline-offset-8 outline-offset-0  rounded-3xl transition-offset duration-200 motion-reduce:duration-0"
    to="pokemon-details/Charizard"
  >
    <div
      class="inner-glow bg-slate-800 inline-flex flex-col justify-center group cursor-pointer  dark:bg-slate-900 p-2 dark:text-slate-50 rounded-3xl dark:border-slate-800 border"
    >
      <div
        class="parallax-effect-glare-scale "
      >
        <figure
          class="popup-element flex items-center justify-center"
        >
          <div
            class="relative w-60 h-60 -mt-16"
          >
            <img
              alt="Charizard image-backdrop"
              class="blur-2xl w-52 absolute opacity-50 group-hover:opacity-90 m-auto inset-0 transition-opacity"
              height="240"
              loading="lazy"
              src="https://example.com/shiny.png"
              width="240"
            />
            <img
              alt="Charizard image"
              class="w-52 m-auto z-10 relative"
              height="240"
              loading="lazy"
              src="https://example.com/shiny.png"
              width="240"
            />
          </div>
        </figure>
        <div
          class="flex items-center justify-center gap-2"
        >
          <h3
            class="text-slate-300 group-hover:text-slate-50 dark:text-slate-400 dark:group-hover:text-slate-50 group-hover:underline capitalize text-center font-semibold my-5 text-xl"
          >
            Charizard
          </h3>
          <svg
            aria-hidden="true"
            aria-label="navigate icon"
            class="lucide lucide-arrow-up-right opacity-0 size-0 scale-0 group-hover:opacity-100 group-hover:scale-100 group-hover:size-auto transition-transform text-slate-50 dark:text-slate-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 7h10v10"
            />
            <path
              d="M7 17 17 7"
            />
          </svg>
        </div>
        <div
          class="glare-wrapper"
          style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; overflow: hidden; border-radius: 24px; pointer-events: none;"
        >
          <div
            class="glare"
            style="position: absolute; top: 50%; left: 50%; transform-origin: 0% 0%; pointer-events: none; width: 0px; height: 0px;"
          />
        </div>
      </div>
    </div>
  </a>
</DocumentFragment>
`;
